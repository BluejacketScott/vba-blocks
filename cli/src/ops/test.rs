pub fn test(filters: Vec<&str>) {
  let filter_value = match filters.len() {
    0 => "(no filters)".to_string(),
    _ => filters.join(" "),
  };

  println!("test, filters: {}", filter_value);
}
