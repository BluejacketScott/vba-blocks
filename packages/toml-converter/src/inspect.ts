import { Path, Index } from 'deep-diff';
import { graceful as detectNewline } from 'detect-newline';
import { Change } from './diff';
import { toLines } from './utils';

// Use __internal__ parsing from toml library
// (this should not be assumed to be stable and a specific version may need to be pinned)
export interface ParsedNode {
  type:
    | 'ObjectPath'
    | 'ArrayPath'
    | 'Assign'
    | 'InlineTable'
    | 'InlineTableValue'
    | 'Array'
    | 'String'
    | 'Float'
    | 'Integer'
    | 'Boolean'
    | 'Date';

  line: number;
  column: number;
  value: any;
  key?: string;
}

const parseToml: (toml: string) => ParsedNode[] = require('toml/lib/parser')
  .parse;

/*
  Approach: Arrange nodes into AST tree and add end information

  Example:
  ```toml
    123456789
  1 [package]
  2 name = "value"
  3 authors = ["a", "b"]
  4
  5 [dependencies.a]
  6 git = { path = "...", tag = "v1.0.0" }
  ```

  ```js
  {
    package: {
      key: {
        start: { line: 1, column: 1 },
        end: { line: 1, column 9 },
        parsed: (from toml = package)
      },
      value: {
        start: { line: 2, column: 1 },
        end: { line: 4, column: 1 },
        parsed: (from toml)
      },
      children: {
        name: {
          key: { start: { ... }, end: { ... }, parsed: (name)},
          value: { start: { ... }, end: { ... }, parsed: (value)}
        },
        authors: {
          key: { ... },
          value: { ... },
          children: [
            { key: null, value: { ..., parsed: (a) } },
            { key: null, value: { ..., parsed: (b) } }
          ]
        }
      }
    },
    dependencies: {
      key: null,
      value: { start: { line: 5 }, ...},
      children: {
        a: {
          key: { ... },
          value: { ... },
          children: {
            git: {
              key: { end: { line: 6, column: 3 } },
              value: { ... },
              children: {
                path: {
                  key: { ... },
                  value: { ..., parsed: (...) }
                },
                tag: {
                  key: { ... },
                  value: { ..., parsed: (v1.0.0) }
                }
              }
            }
          }
        }
      }
    }
  }
  ```
*/

export type AST = KeyedNodes;

export interface KeyedNodes {
  [key: string]: Node;
}
export type IndexedNodes = Node[];

export interface Node {
  key?: Value;
  value: Value;
  children?: KeyedNodes | IndexedNodes;
}

export interface Value {
  start: Position;
  end: Position;
  parsed?: ParsedNode;
}

export interface Position {
  line: number;
  column: number;
}

export default function inspect(toml: string): AST {
  const lines = toLines(toml);
  const nodes: ParsedNode[] = parseToml(lines.join('\n'));
  return <KeyedNodes>walk(nodes, lines);
}

function walk(parsed: ParsedNode[], lines: string[]): KeyedNodes {
  let nodes: KeyedNodes = {};
  let current: KeyedNodes = nodes;

  for (const [index, node] of parsed.entries()) {
    let { line, column } = node;
    let key: null | Value;
    let value: Value;
    let children: KeyedNodes | IndexedNodes;
    let result: Node;

    let next_node: ParsedNode | undefined;
    let end_line: number;
    let end_of_key: number;
    let end_of_value: number;

    switch (node.type) {
      case 'ObjectPath':
      case 'ArrayPath':
        // Parsed ArrayPath points to first line of values (not key)
        // if (node.type === 'ArrayPath') line -= 1;

        // { "value": ["package"] }
        next_node = parsed.find(
          (node, _index) => _index > index && isPathNode(node)
        );
        end_line = next_node ? next_node.line - 1 : lines.length;

        // TODO something is wrong with end.column position (-> line length)

        result = {
          key: {
            start: { line, column },
            end: { line, column: lines[line - 1].length },
            parsed: node
          },
          value: {
            start: { line: line + 1, column: 0 },
            end: { line: end_line, column: lines[end_line - 1].length }
          },
          children: {}
        };

        current = <KeyedNodes>result.children;

        node.type === 'ObjectPath'
          ? set(nodes, node.value, result)
          : push(nodes, node.value, result);

        break;

      case 'Assign':
        // { "key": "name", "value": { "type": "String", "value": "package-name" }
        next_node = parsed[index + 1];
        end_line = next_node ? next_node.line - 1 : lines.length;
        end_of_key = findCharColumn(lines[line - 1], '=', {
          from: column,
          fallback: column + node.key!.length
        });

        result = {
          key: {
            start: { line, column },
            end: { line, column: end_of_key },
            parsed: node
          },
          value: {
            start: { line, column: end_of_key + 1 },
            end: { line: end_line, column: lines[end_line - 1].length },
            parsed: node.value
          },
          children: toChildren(node.value, lines)
        };

        set(current, [node.key!], result);

        break;

      case 'InlineTableValue':
        // { "key": "name", "value": { "type": "String", "value": "package-name" }
        end_of_key = findCharColumn(lines[line - 1], '=', {
          from: column,
          fallback: column + node.key!.length
        });
        end_of_value = findCharColumn(lines[line - 1], [',', '}'], {
          from: end_of_key
        });

        result = {
          key: {
            start: { line, column },
            end: { line, column: end_of_key },
            parsed: node
          },
          value: {
            start: { line, column: end_of_key + 1 },
            end: { line, column: end_of_value },
            parsed: node.value
          },
          children: toChildren(node.value, lines)
        };

        set(current, [node.key!], result);

        break;

      default:
        throw new Error(
          `Unrecognized/unhandled parsed node with type "${node.type}"`
        );
    }
  }

  return nodes;
}

function toChildren(
  value: ParsedNode,
  lines: string[]
): KeyedNodes | IndexedNodes | undefined {
  switch (value.type) {
    case 'InlineTable':
      return walk(value.value, lines);
    case 'Array':
      return <IndexedNodes>value.value.map((child: ParsedNode) => {
        const { line, column } = child;
        const children = toChildren(child, lines);

        const end_line = Array.isArray(children)
          ? children[children.length - 1].value.end.line
          : line;
        const search_column = Array.isArray(children)
          ? children[children.length - 1].value.end.column
          : column;
        const end_of_value = findCharColumn(lines[end_line - 1], ',', {
          from: search_column
        });

        const value: Value = {
          start: { line, column },
          end: { line, column: end_of_value },
          parsed: child
        };

        return { value, children };
      });
    default:
      return;
  }
}

export function get(ast: AST, path: Path): Node | null {
  if (!path.length) return null;

  const [root, ...nested] = path;
  let node: Node = ast[root];

  for (const part of nested) {
    const children = node && node.children;
    if (!children) return null;

    node = Array.isArray(children)
      ? children[<number>part]
      : children[<string>part];
  }

  return node;
}

export function has(ast: AST, path: Path): boolean {
  return !!get(ast, path);
}

function set(ast: AST, path: Path, node: Node): Node {
  if (path.length === 1) {
    return (ast[path[0]] = node);
  }

  const key: string = <string>path[path.length - 1];
  const parent_path = path.slice(0, -1);
  let parent = get(ast, parent_path);

  if (!parent) {
    parent = {
      value: {
        start: node.key!.start,
        end: node.value.end
      },
      children: {}
    };

    set(ast, parent_path, parent);
  }
  if (!parent.children) {
    parent.children = {};
  }

  // Each time a new value is add to parent,
  // Ensure end location is up-to-date
  if (node.value.end.line > parent.value.end.line) {
    parent.value.end = node.value.end;
  }

  return ((<KeyedNodes>parent.children)[key] = node);
}

function push(ast: AST, path: Path, node: Node): Node {
  let parent = get(ast, path);

  if (!parent) {
    parent = {
      value: {
        start: node.key!.start,
        end: node.value.end
      },
      children: []
    };

    set(ast, path, parent);
  }
  if (!parent.children) {
    parent.children = [];
  }

  // Each time a new value is pushed onto path,
  // update the end position of the parent's value
  parent.value.end = node.value.end;
  (<IndexedNodes>parent.children!).push(node);

  return node;
}

function isPathNode(node: ParsedNode): boolean {
  return node.type === 'ObjectPath' || node.type === 'ArrayPath';
}

const QUOTE = '"';
const ESCAPE = '\\';

function findCharColumn(
  value: string,
  search: string | string[],
  options?: { from?: number; fallback?: number }
): number {
  const { from = 1, fallback = value.length } = options || {};

  if (!Array.isArray(search)) search = [search];
  for (const check of search) {
    let quoted = false;
    for (let i = from; i <= value.length; i++) {
      const char = value.charAt(i - 1);
      if (char === check && !quoted) return i;

      if (!quoted && char === QUOTE) quoted = true;
      if (quoted && char === QUOTE && value.charAt(i - 2) !== ESCAPE)
        quoted = false;
    }
  }

  return fallback;
}
