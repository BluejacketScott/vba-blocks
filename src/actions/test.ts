import env from '../env';
import run from './run';

export interface TestOptions {
  target?: string;
}

export default async function test(options: TestOptions) {
  const { target } = options;
  const macro = 'Tests.Run';
  const stdout = env.isWindows ? 'CON' : '/dev/stdout';

  await run({ target, macro, args: [stdout] });
}
