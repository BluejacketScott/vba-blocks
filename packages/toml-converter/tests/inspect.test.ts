import inspect from '../src/inspect';
import { fixture } from './patch.test';

test('inspect', () => {
  expect(inspect(fixture)).toMatchSnapshot();
});
