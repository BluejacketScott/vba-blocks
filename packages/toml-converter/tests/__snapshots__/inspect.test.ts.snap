// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`inspect 1`] = `
Object {
  "array": Object {
    "children": Array [
      Object {
        "children": Object {
          "index": Object {
            "children": undefined,
            "key": Object {
              "end": Object {
                "column": 7,
                "line": 23,
              },
              "parsed": Object {
                "column": 1,
                "key": "index",
                "line": 23,
                "type": "Assign",
                "value": Object {
                  "column": 9,
                  "line": 23,
                  "type": "Integer",
                  "value": 0,
                },
              },
              "start": Object {
                "column": 1,
                "line": 23,
              },
            },
            "value": Object {
              "end": Object {
                "column": 0,
                "line": 24,
              },
              "parsed": Object {
                "column": 9,
                "line": 23,
                "type": "Integer",
                "value": 0,
              },
              "start": Object {
                "column": 8,
                "line": 23,
              },
            },
          },
        },
        "key": Object {
          "end": Object {
            "column": 14,
            "line": 22,
          },
          "parsed": Object {
            "column": 1,
            "line": 22,
            "type": "ArrayPath",
            "value": Array [
              "array",
            ],
          },
          "start": Object {
            "column": 1,
            "line": 22,
          },
        },
        "value": Object {
          "end": Object {
            "column": 0,
            "line": 24,
          },
          "start": Object {
            "column": 0,
            "line": 23,
          },
        },
      },
      Object {
        "children": Object {
          "index": Object {
            "children": undefined,
            "key": Object {
              "end": Object {
                "column": 7,
                "line": 26,
              },
              "parsed": Object {
                "column": 1,
                "key": "index",
                "line": 26,
                "type": "Assign",
                "value": Object {
                  "column": 9,
                  "line": 26,
                  "type": "Integer",
                  "value": 1,
                },
              },
              "start": Object {
                "column": 1,
                "line": 26,
              },
            },
            "value": Object {
              "end": Object {
                "column": 0,
                "line": 27,
              },
              "parsed": Object {
                "column": 9,
                "line": 26,
                "type": "Integer",
                "value": 1,
              },
              "start": Object {
                "column": 8,
                "line": 26,
              },
            },
          },
        },
        "key": Object {
          "end": Object {
            "column": 9,
            "line": 25,
          },
          "parsed": Object {
            "column": 1,
            "line": 25,
            "type": "ArrayPath",
            "value": Array [
              "array",
            ],
          },
          "start": Object {
            "column": 1,
            "line": 25,
          },
        },
        "value": Object {
          "end": Object {
            "column": 0,
            "line": 27,
          },
          "start": Object {
            "column": 0,
            "line": 26,
          },
        },
      },
      Object {
        "children": Object {
          "index": Object {
            "children": undefined,
            "key": Object {
              "end": Object {
                "column": 7,
                "line": 29,
              },
              "parsed": Object {
                "column": 1,
                "key": "index",
                "line": 29,
                "type": "Assign",
                "value": Object {
                  "column": 9,
                  "line": 29,
                  "type": "Integer",
                  "value": 2,
                },
              },
              "start": Object {
                "column": 1,
                "line": 29,
              },
            },
            "value": Object {
              "end": Object {
                "column": 9,
                "line": 29,
              },
              "parsed": Object {
                "column": 9,
                "line": 29,
                "type": "Integer",
                "value": 2,
              },
              "start": Object {
                "column": 8,
                "line": 29,
              },
            },
          },
        },
        "key": Object {
          "end": Object {
            "column": 9,
            "line": 28,
          },
          "parsed": Object {
            "column": 1,
            "line": 28,
            "type": "ArrayPath",
            "value": Array [
              "array",
            ],
          },
          "start": Object {
            "column": 1,
            "line": 28,
          },
        },
        "value": Object {
          "end": Object {
            "column": 9,
            "line": 29,
          },
          "start": Object {
            "column": 0,
            "line": 29,
          },
        },
      },
    ],
    "value": Object {
      "end": Object {
        "column": 9,
        "line": 29,
      },
      "start": Object {
        "column": 1,
        "line": 22,
      },
    },
  },
  "dependencies": Object {
    "children": Object {
      "a": Object {
        "children": undefined,
        "key": Object {
          "end": Object {
            "column": 3,
            "line": 13,
          },
          "parsed": Object {
            "column": 1,
            "key": "a",
            "line": 13,
            "type": "Assign",
            "value": Object {
              "column": 5,
              "line": 13,
              "type": "String",
              "value": "1.0.0",
            },
          },
          "start": Object {
            "column": 1,
            "line": 13,
          },
        },
        "value": Object {
          "end": Object {
            "column": 11,
            "line": 13,
          },
          "parsed": Object {
            "column": 5,
            "line": 13,
            "type": "String",
            "value": "1.0.0",
          },
          "start": Object {
            "column": 4,
            "line": 13,
          },
        },
      },
      "b": Object {
        "children": Object {
          "git": Object {
            "children": undefined,
            "key": Object {
              "end": Object {
                "column": 11,
                "line": 14,
              },
              "parsed": Object {
                "column": 7,
                "key": "git",
                "line": 14,
                "type": "InlineTableValue",
                "value": Object {
                  "column": 13,
                  "line": 14,
                  "type": "String",
                  "value": "...git",
                },
              },
              "start": Object {
                "column": 7,
                "line": 14,
              },
            },
            "value": Object {
              "end": Object {
                "column": 23,
                "line": 14,
              },
              "parsed": Object {
                "column": 13,
                "line": 14,
                "type": "String",
                "value": "...git",
              },
              "start": Object {
                "column": 12,
                "line": 14,
              },
            },
          },
          "spacing": Object {
            "children": undefined,
            "key": Object {
              "end": Object {
                "column": 33,
                "line": 14,
              },
              "parsed": Object {
                "column": 25,
                "key": "spacing",
                "line": 14,
                "type": "InlineTableValue",
                "value": Object {
                  "column": 38,
                  "line": 14,
                  "type": "String",
                  "value": "weird",
                },
              },
              "start": Object {
                "column": 25,
                "line": 14,
              },
            },
            "value": Object {
              "end": Object {
                "column": 45,
                "line": 14,
              },
              "parsed": Object {
                "column": 38,
                "line": 14,
                "type": "String",
                "value": "weird",
              },
              "start": Object {
                "column": 34,
                "line": 14,
              },
            },
          },
        },
        "key": Object {
          "end": Object {
            "column": 3,
            "line": 14,
          },
          "parsed": Object {
            "column": 1,
            "key": "b",
            "line": 14,
            "type": "Assign",
            "value": Object {
              "column": 5,
              "line": 14,
              "type": "InlineTable",
              "value": Array [
                Object {
                  "column": 7,
                  "key": "git",
                  "line": 14,
                  "type": "InlineTableValue",
                  "value": Object {
                    "column": 13,
                    "line": 14,
                    "type": "String",
                    "value": "...git",
                  },
                },
                Object {
                  "column": 25,
                  "key": "spacing",
                  "line": 14,
                  "type": "InlineTableValue",
                  "value": Object {
                    "column": 38,
                    "line": 14,
                    "type": "String",
                    "value": "weird",
                  },
                },
              ],
            },
          },
          "start": Object {
            "column": 1,
            "line": 14,
          },
        },
        "value": Object {
          "end": Object {
            "column": 0,
            "line": 15,
          },
          "parsed": Object {
            "column": 5,
            "line": 14,
            "type": "InlineTable",
            "value": Array [
              Object {
                "column": 7,
                "key": "git",
                "line": 14,
                "type": "InlineTableValue",
                "value": Object {
                  "column": 13,
                  "line": 14,
                  "type": "String",
                  "value": "...git",
                },
              },
              Object {
                "column": 25,
                "key": "spacing",
                "line": 14,
                "type": "InlineTableValue",
                "value": Object {
                  "column": 38,
                  "line": 14,
                  "type": "String",
                  "value": "weird",
                },
              },
            ],
          },
          "start": Object {
            "column": 4,
            "line": 14,
          },
        },
      },
      "c": Object {
        "children": Object {
          "path": Object {
            "children": undefined,
            "key": Object {
              "end": Object {
                "column": 6,
                "line": 17,
              },
              "parsed": Object {
                "column": 1,
                "key": "path",
                "line": 17,
                "type": "Assign",
                "value": Object {
                  "column": 8,
                  "line": 17,
                  "type": "String",
                  "value": "...path",
                },
              },
              "start": Object {
                "column": 1,
                "line": 17,
              },
            },
            "value": Object {
              "end": Object {
                "column": 0,
                "line": 18,
              },
              "parsed": Object {
                "column": 8,
                "line": 17,
                "type": "String",
                "value": "...path",
              },
              "start": Object {
                "column": 7,
                "line": 17,
              },
            },
          },
        },
        "key": Object {
          "end": Object {
            "column": 20,
            "line": 16,
          },
          "parsed": Object {
            "column": 1,
            "line": 16,
            "type": "ObjectPath",
            "value": Array [
              "dependencies",
              "c",
            ],
          },
          "start": Object {
            "column": 1,
            "line": 16,
          },
        },
        "value": Object {
          "end": Object {
            "column": 0,
            "line": 18,
          },
          "start": Object {
            "column": 0,
            "line": 17,
          },
        },
      },
    },
    "key": Object {
      "end": Object {
        "column": 19,
        "line": 12,
      },
      "parsed": Object {
        "column": 1,
        "line": 12,
        "type": "ObjectPath",
        "value": Array [
          "dependencies",
        ],
      },
      "start": Object {
        "column": 1,
        "line": 12,
      },
    },
    "value": Object {
      "end": Object {
        "column": 0,
        "line": 18,
      },
      "start": Object {
        "column": 0,
        "line": 13,
      },
    },
  },
  "package": Object {
    "children": Object {
      "authors": Object {
        "children": Array [
          Object {
            "children": undefined,
            "value": Object {
              "end": Object {
                "column": 22,
                "line": 4,
              },
              "parsed": Object {
                "column": 12,
                "line": 4,
                "type": "String",
                "value": "Author 1",
              },
              "start": Object {
                "column": 12,
                "line": 4,
              },
            },
          },
        ],
        "key": Object {
          "end": Object {
            "column": 9,
            "line": 4,
          },
          "parsed": Object {
            "column": 1,
            "key": "authors",
            "line": 4,
            "type": "Assign",
            "value": Object {
              "column": 11,
              "line": 4,
              "type": "Array",
              "value": Array [
                Object {
                  "column": 12,
                  "line": 4,
                  "type": "String",
                  "value": "Author 1",
                },
              ],
            },
          },
          "start": Object {
            "column": 1,
            "line": 4,
          },
        },
        "value": Object {
          "end": Object {
            "column": 0,
            "line": 11,
          },
          "parsed": Object {
            "column": 11,
            "line": 4,
            "type": "Array",
            "value": Array [
              Object {
                "column": 12,
                "line": 4,
                "type": "String",
                "value": "Author 1",
              },
            ],
          },
          "start": Object {
            "column": 10,
            "line": 4,
          },
        },
      },
      "name": Object {
        "children": undefined,
        "key": Object {
          "end": Object {
            "column": 6,
            "line": 2,
          },
          "parsed": Object {
            "column": 1,
            "key": "name",
            "line": 2,
            "type": "Assign",
            "value": Object {
              "column": 8,
              "line": 2,
              "type": "String",
              "value": "package-name",
            },
          },
          "start": Object {
            "column": 1,
            "line": 2,
          },
        },
        "value": Object {
          "end": Object {
            "column": 31,
            "line": 2,
          },
          "parsed": Object {
            "column": 8,
            "line": 2,
            "type": "String",
            "value": "package-name",
          },
          "start": Object {
            "column": 7,
            "line": 2,
          },
        },
      },
      "version": Object {
        "children": undefined,
        "key": Object {
          "end": Object {
            "column": 9,
            "line": 3,
          },
          "parsed": Object {
            "column": 1,
            "key": "version",
            "line": 3,
            "type": "Assign",
            "value": Object {
              "column": 11,
              "line": 3,
              "type": "String",
              "value": "0.0.0",
            },
          },
          "start": Object {
            "column": 1,
            "line": 3,
          },
        },
        "value": Object {
          "end": Object {
            "column": 17,
            "line": 3,
          },
          "parsed": Object {
            "column": 11,
            "line": 3,
            "type": "String",
            "value": "0.0.0",
          },
          "start": Object {
            "column": 10,
            "line": 3,
          },
        },
      },
    },
    "key": Object {
      "end": Object {
        "column": 9,
        "line": 1,
      },
      "parsed": Object {
        "column": 1,
        "line": 1,
        "type": "ObjectPath",
        "value": Array [
          "package",
        ],
      },
      "start": Object {
        "column": 1,
        "line": 1,
      },
    },
    "value": Object {
      "end": Object {
        "column": 0,
        "line": 11,
      },
      "start": Object {
        "column": 0,
        "line": 2,
      },
    },
  },
  "references": Object {
    "children": Object {
      "nested": Object {
        "children": Object {
          "guid": Object {
            "children": undefined,
            "key": Object {
              "end": Object {
                "column": 6,
                "line": 20,
              },
              "parsed": Object {
                "column": 1,
                "key": "guid",
                "line": 20,
                "type": "Assign",
                "value": Object {
                  "column": 8,
                  "line": 20,
                  "type": "String",
                  "value": "{...}",
                },
              },
              "start": Object {
                "column": 1,
                "line": 20,
              },
            },
            "value": Object {
              "end": Object {
                "column": 0,
                "line": 21,
              },
              "parsed": Object {
                "column": 8,
                "line": 20,
                "type": "String",
                "value": "{...}",
              },
              "start": Object {
                "column": 7,
                "line": 20,
              },
            },
          },
        },
        "key": Object {
          "end": Object {
            "column": 23,
            "line": 19,
          },
          "parsed": Object {
            "column": 1,
            "line": 19,
            "type": "ObjectPath",
            "value": Array [
              "references",
              "nested",
            ],
          },
          "start": Object {
            "column": 1,
            "line": 19,
          },
        },
        "value": Object {
          "end": Object {
            "column": 0,
            "line": 21,
          },
          "start": Object {
            "column": 0,
            "line": 20,
          },
        },
      },
    },
    "value": Object {
      "end": Object {
        "column": 0,
        "line": 21,
      },
      "start": Object {
        "column": 1,
        "line": 19,
      },
    },
  },
}
`;

exports[`patch 1`] = `
"[package]
name = \\"package-name\\" # comment
version = \\"0.0.0\\"
authors = [\\"Author 1\\"]

# 6
# 7
# 8

# 10

[dependencies] # 12
a = \\"1.0.0\\"
b = { git = \\"...git\\"  , spacing =    \\"weird\\"}#comment

[dependencies.c] #16
path = \\"...path\\"

[references.nested] #19
guid = \\"{...}\\"

[[array]] # 22
index = 0

[[array]]
index = 1

[[array]]
index = 2"
`;
